File: iOS/IssueTracker/IssueTracker/Model/Issue.swift
//
//  Issue.swift
//  IssueTracker
//
//  Created by 양어진 on 2020/11/10.
//  Copyright © 2020 양어진. All rights reserved.
//

import Foundation

struct Issue: Codable {
    let id: Int?
    let title: String?
    let contents: String?
    let status: Int?
    let author: User?
    let authorID: Int?
    let milestone: Milestone?
    let milestoneID: Int?
    let assignees: [User]?
    let labels: [Label]?
    let comments: [Comment]?
    let createdAt: String?
    let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case title, contents, status, author, labels, assignees, milestone, authorID, milestoneID, comments
        case id = "issue_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

extension Issue: Hashable {
    
    static func == (lhs: Issue, rhs: Issue) -> Bool {
        return lhs.id == rhs.id
            && lhs.title == rhs.title
            && lhs.contents == rhs.contents
            && lhs.status == rhs.status
            && lhs.author == rhs.author
            && lhs.assignees == rhs.assignees
            && lhs.milestone == rhs.milestone
            && lhs.comments == rhs.comments
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
        hasher.combine(authorID)
        hasher.combine(title)
        hasher.combine(contents)
        hasher.combine(status)
        hasher.combine(labels)
        hasher.combine(author)
        hasher.combine(assignees)
        hasher.combine(milestone)
        hasher.combine(milestoneID)
        hasher.combine(comments)
    }
    
}

File: web/server/objects/issue.js
import Label from "./label.js";
import Milestone from "./milestone.js";
import User from "./user.js";

export default class Issue {
  constructor(item) {
    (this.issue_id = item.issue_id),
      (this.title = item.title),
      (this.contents = item.contents),
      (this.status = item.status),
      (this.author = new User(item.author, item.author_nickname)),
      (this.assignees = []),
      (this.labels = []),
      (this.milestone = new Milestone(
        item.milestone_id,
        item.milestone_title,
        item.milestone_status,
        item.milestone_issuesStatus
      )),
      (this.created_at = item.created_at),
      (this.updated_at = item.updated_at),
      (this.deleted_at = item.deleted_at);

    if (item.label_id !== null) this.setLabel(item);
    if (item.assignee_id !== null) this.setAssignee(item);
  }

  setLabel = (item) => {
    let hasLabel = false;
    this.labels.forEach((label) => {
      if (label.label_id === item.label_id) {
        hasLabel = true;
        return;
      }
    });
    if (hasLabel === false)
      this.labels.push(
        new Label(item.label_id, item.label_name, item.label_color)
      );
  };

  setAssignee = (item) => {
    let hasAssignee = false;
    this.assignees.forEach((assignee) => {
      if (assignee.user_id === item.assignee_id) {
        hasAssignee = true;
        return;
      }
    });
    if (hasAssignee === false)
      this.assignees.push(new User(item.assignee_id, item.assignee_nickname));
  };
}
